scripts: twister: drop support for space-separated lists

gmarull opened the pull request:
Support for space-separated lists was deprecated a long time ago
https://github.com/zephyrproject-rtos/zephyr/commit/a91620f5bbf2a01c5cdc4b435e8b00cc6020be11 so it is time to remove
support for them.

Any project that has not migrated can still use
`scripts/utils/twister_to_list.py` to automatically migrate twister
files.

gmarull committed:
scripts: twister: drop support for space-separated lists

Support for space-separated lists was deprecated a long time ago
a91620f5bbf2a01c5cdc4b435e8b00cc6020be11 so it is time to remove
support for them.

Any project that has not migrated can still use
`scripts/utils/twister_to_list.py` to automatically migrate twister
files.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
doc: rel-notes: inform about space-separated list removal in Twister

Inform about the removal of space-separated lists in Twister config
files. Also provide guidance on how to migrated using a utility
script.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gchwier commented on scripts/pylib/twister/twisterlib/config_parser.py:
```diff
@@ -124,39 +124,15 @@ def _cast_value(self, value, typestr):
-                    return [self._cast_value(vsi, typestr[5:]) for vsi in vs]
-                else:
-                    return vs
+                return [value]
```
```suggestion
                return [value] if value else []
```

gmarull replied:
thanks for spotting the issue!

gchwier commented on scripts/pylib/twister/twisterlib/config_parser.py:
```diff
@@ -124,39 +124,15 @@ def _cast_value(self, value, typestr):
-                    return {self._cast_value(vsi, typestr[4:]) for vsi in vs}
-                else:
-                    return set(vs)
+                return {value}
```
```suggestion
                return set(value) if value else set()
```

gmarull committed:
samples, tests: remove usage of space-separated lists

Convert them to native YAML lists. Support for space-separated
lists was deprecated in Twister a long time ago.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gchwier reviewed (commented):
tests failed, because the default values are set to {''} and [''],
e.g. in blackbox tests the required_snippets is not set, and twister tries to process ['']:
`ERROR    twister:testplan.py:1062 Can't find snippet '' for test ...`

gchwier commented:
Think also to update https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/schemas/twister/testsuite-schema.yaml
change:
```
      type: any
      required: false
```
to 
```
      type: seq
      required: false
      sequence:
        - type: str
```

gmarull commented:
last push: some fixes to the config parser, where string-based lists were still constructed

fabiobaltieri reviewed (approved):
I'm not a robot. Beep. 🤖

henrikbrixandersen commented:
> Support for space-separated lists was deprecated a long time ago

Perhaps so, but as discussed in Discord (https://discord.com/channels/720317445772017664/733037890514321419/1314192033484902470) this deprecation never resulted in any CI warnings/failures where the deprecated list form was used - which is also evident by the fact that this PR needed to remove the remaining uses of space-separated lists in-tree.

This goes against our usual process for deprecation in Zephyr, and I'm sure this will hit (even more) downstream users once this change propagates. Downstream users will have had no indication that they were using deprecated syntax if they had not explicitly enabled `PYTHONWARNINGS` in their builds/CI (which we seemingly do not enable in our upstream infrastructure either).

This is very unfortunate.

henrikbrixandersen commented:
> Change may be reverted for one more release then, but making sure upstream fails if feature is used.

I think this is the best approach.
